iphone/Maps/UI/PlacePage/BannersCache.swift

```diff
@@ -0,0 +1,81 @@
+import FBAudienceNetwork
+
+@objc (MWMBannersCache)
+final class BannersCache: NSObject {
+
+  fileprivate var cache: [String : FBNativeAd] = [:]
+  fileprivate var requests: [String] = []
+  fileprivate let worker = DispatchQueue(label: "com.mapsme.banners", qos: .userInteractive)
+
+  typealias Completion = (FBNativeAd) -> Void
+  fileprivate var completion: Completion?
+
+  func get(_ key: String, completion comp: Completion?) -> FBNativeAd? {
+    completion = comp
+    if requests.index(of: key) != nil {
+      return cache[key];
+    }
+
+    worker.async {
+      let banner = FBNativeAd(placementID: key)
+      banner.mediaCachePolicy = .all
+      banner.delegate = self
+      banner.load()
+      self.requests.append(key)
+    }
+
+    return cache[key];
+  }
+
+  fileprivate func put(_ banner: FBNativeAd) {
+    cache[banner.placementID] = banner;
+  }
+
+  fileprivate func removeRequest(atKey key: String) {
+    if let index = requests.index(of: key) {
+      requests.remove(at: index)
+    }
+    else {
+      assert(false)
+    }
+  }
+
+}
+
+extension BannersCache : FBNativeAdDelegate {
+
+  func nativeAdDidLoad(_ nativeAd: FBNativeAd) {
+    let key = nativeAd.placementID
+
+    Statistics.logEvent(kStatPlacePageBannerShow, withParameters: [kStatBanner : key, kStatProvider : kStatFacebook])
+
+    put(nativeAd)
+    removeRequest(atKey: key)
+    DispatchQueue.main.async {
+      self.completion?(nativeAd)
+    }
+  }
+
+  func nativeAd(_ nativeAd: FBNativeAd, didFailWithError error: Error) {
+    let key = nativeAd.placementID;
+
+    // https://developers.facebook.com/docs/audience-network/testing
+    var params: [String: Any] = [kStatBanner : key, kStatProvider : kStatFacebook]
+    let e = error as NSError
```

Comment:
Кто бы мог подумать:
The reason this works is because of "compiler magic." The compiler automatically emits the code necessary to translate between any ErrorType and NSError.